scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -1307,17 +1320,42 @@ def __select_processes(self):
                 self.__pids = [self.__pids[0]]
         else:
             # See if the specified target pid is running.  If so, then return it.
-            # Special case '$$' to mean this process.
+            # Special cases:
+            #   '$$' mean this process.
+            #   '$$TBD' mean that the PID of the target process has not been determined yet and it will be set later.
             pids = []
-            if self.__target_pids:
-                for t_pid in self.__target_pids:
+            with self._target_pids_lock:
+                target_pids = self.__target_pids[:]
+            if target_pids:
+                for t_pid in target_pids:
                     if t_pid == "$$":
                         t_pid = int(os.getpid())
+
+                    # skip this until it will be replaced with a real PID.
+                    elif t_pid == "$$TBD":
+                        continue
                     else:
                         t_pid = int(t_pid)
                     pids.append(t_pid)
             self.__pids = pids
         return self.__pids
 
+    def set_pid(self, pid):  # type: (int) -> None
+        """
+        Set the PID of the process that was marked as $$TBD.
+        NOTE: this method is called from the a different thread, so all shared objects must be guarded.
+        :param pid: Process PID
+        """
+        with self._target_pids_lock:
+            target_pids = self.__target_pids[:]
+        for i in range(len(target_pids)):
+            if target_pids[i] == "$$TBD":
```

Comment:
So it seems this sets the pid on the first "$$TBD" value it sees, right?

How is the logic supposed to work? We set id in local ``target_pids`` list variable we create (which is a copy of ``self.__target_pids``, but local to this function) and we never update ``self.__target_pids`` itself.

Or am I missing something?

And if we will update ``self.__target_pids`` as part of this function, we will also need a read + write lock here since it can be called by multiple workers at the same time, right?

Reply:
Oh right, I don't assign the temporary list back to `self.__target_pids`.  But anyway, as you mentioned above, it should be ok to not use the lock at all, since we only write the PID once for each monitor and the write itself should be atomic.